src/test/java/org/cloudfoundry/promregator/rewrite/MetricFamilySamplesEnricherTest.java

```diff
@@ -104,4 +104,48 @@ public void testNullEnricherSimple() {
 		Assert.assertEquals("labelValue", labelValues[0]);
 	}
 
+	@Test
+	public void testDuplicatesRemovedFromMetricLabels() {
+		AbstractMetricFamilySamplesEnricher subject = new CFAllLabelsMetricFamilySamplesEnricher("testOrgName", "testSpaceName", "testComponent", "testInstance:42");
+
+		List<Sample> samples = new LinkedList<>();
+		Sample s = new Sample("dummyname", Arrays.asList(new String[] { "app_name", "labelName" }), Arrays.asList(new String[] { "original_appName", "labelvalue"}), 1.0);
+		samples.add(s);
+
+		MetricFamilySamples mfs = new MetricFamilySamples("dummyname", Type.GAUGE, "dummyHelp", samples);
+
+		HashMap<String, MetricFamilySamples> map = new HashMap<>();
+		map.put("metricName", mfs);
+
+		HashMap<String, MetricFamilySamples> result = subject.determineEnumerationOfMetricFamilySamples(map);
+
+		Assert.assertEquals(1, result.size());
+
+		MetricFamilySamples testMFS = result.get("metricName");
+		Assert.assertNotNull(testMFS);
+
+		Sample testSample = testMFS.samples.get(0);
+		Assert.assertNotNull(testSample);
+
+		List<String> labelNamesList = testSample.labelNames;
+		Assert.assertEquals("There should be exactly six labels", 6,labelNamesList.size());
+		String[] labelNames = labelNamesList.toArray(new String[0]);
+		Assert.assertEquals("The first label should have been the app_name", "app_name", labelNames[0]);
+		Assert.assertEquals("The first label should have been the labelName", "labelName", labelNames[1]);
+		Assert.assertEquals("The first label should have been the testOrgName", "org_name", labelNames[2]);
+		Assert.assertEquals("The first label should have been the testSpaceName", "space_name", labelNames[3]);
+		Assert.assertEquals("The first label should have been the testComponent", "cf_instance_id", labelNames[4]);
+		Assert.assertEquals("The first label should have been the testComponent", "cf_instance_number", labelNames[5]);
+
+		List<String> labelValuesList = testSample.labelValues;
+		Assert.assertEquals("There should be exactly 6 values",6, labelValuesList.size());
+		String[] labelValues = labelValuesList.toArray(new String[0]);
+		Assert.assertEquals("The app_name should not have been empty or overridden", "original_appName", labelValues[0]);
```

Comment:
I actually would prefer the app_name be overridden by Promregator, so an app isn't in the place too 'spoof' another app's identity/metrics.

Reply:
That's also a valid point....

@mcsammac How does it look like in your practice: Is the data provided by the apps valid (and hence just redundant) or are these guys "little pretenders" 😁 ?

Reply:
The names are valid in a since that they represent the app, but it is not the same as the app name in PCF.  I originally was going to go down the route of giving the enriched labels name precedent, but I ultimately thought it would be better to give devs the ability to override if they saw a need too.

Reply:
Sounds good to me

Reply:
You see me a little astonished. 
For the `appName` I still could get your way of thinking -- if they want to call their app `donkey` in their reporting instead, then fine... However, how do you want to make sure then that they don't overwrite the `orgName` and `spaceName` as well? By that they could create a huge mess not just at their own app, but also at their neighbor's. Or do you not have any "multi-tenancy"-like environment where this would play a role?

Reply:
The more I think of this case, the more I am asking myself why they are overwriting `appName` in first place. Do they need to retain compatibility? Don't they like what they see? Did they not know that this was auto-added for them? Do we collide by chance with some automatic extractor they use? Are they able (but unwilling only) to choose a their own, custom label name or can't they change?

Do you have contact with some of these folks such that you could ask?

Reply:
In the particular case of our devs it was not intentional, but i could see a developer making a case for wanting to modify it.  Particularly for simplification of prometheus queries. I can also see the case for protecting devs from themselves.  Could do a configuration toggle here as well that is toggled to be not overridden by default, but that may be a bit overboard.  Thoughts?

Reply:
I would recommend against letting them override it, so not even offering a config option. If a user had an actual need to alter a specific label, Prometheus relabeling would suffice so we don't need to implement anything to support even that usecase.

Reply:
I initially also chewed on the toggle idea, but I fear that this could become a configuration nightmare. What if there are "some devs" out there who need to protection from themselves - and some others who "know what they are doing"... ? :dizzy_face:

> Particularly for simplification of prometheus queries

Do you see a reason why they can't be simplified using another (custom) label name for the value?

> Thoughts?

I am currently on the trip to forbid it (i.e. overwriting the label values). Almost any time I can think of, we are running in some risk here. 
Too bad that we didn't prefix all labels by Promregator initially with an own namespace (e.g. `promregator_appName` or so). That would have had the implicit tag attached "Keep your fingers off"... 😭 


Side note/internal remark: This all happens only in case label enrichment is used. We don't have such problems with label rewriting (at Prometheus), as the label name there is configured by the user.  ==> won't be a topic after "Promregator-simplification" anymore.

Reply:
One more thing to consider: If we forbid it, what would be the section that we write this change to in the Changelog (when we released the next version with it)?
1. Prometheus de facto did an incompatible change in a patch (as their impact mainly was bug-like) for which we need to adapt now.
2. We would suppressing data from the applications and hence were also incompatible. 
3. The protection preventing to overwrite could also be sold as an improvement/enhancement. 

Talking about [semver](https://semver.org/), item 2 would call out for a major bump. Item 1  hints more towards a patch version increment. And item 3 speaks for a new minor version. 😵

Reply:
That was also something i considered (changing the label values), but was afraid of having to reconfigure a bunch of dashboards.  I'm good with protecting the developers and the ones that have a specific use case using the prometheus relabeling.  So i purpose this:  

1. Check for any collisons with the enriched labels in getEnrichedLabelNames not adding the colliding label/s to the label list.
2. Doing the same check in getEnrichedLabelValues and not adding the corresponding value/s. 

Should it still log a warning if there is a collison, if it is noted as a feature?  And if so do we only log in one of the checks?

Reply:
> what would be the section that we write this change to in the Changelog

A minor bump would call attention to the change, but since it was bug like anyways, it shouldn't have been being used. So I would be fine with minor or patch.

Reply:
> Should it still log a warning if there is a collision

This should be a rare situation. So logging each time as a warning would be fine. Make it a sub-logger so it can be muted separately from the main logger.

Out of curiousity, why do you use label enrichment instead of Prometheus relabeling?

Reply:
We have been using Promregator for nearly a year, and honestly we were running a very old version until very recently, when we had a team that had capacity to own it.  When was relabeling introduced? I don't know that it was in place when we first started using Promregator, or we were being lazy 😄

Reply:
It was introduced with 5ecf6e5def2482c56d43f47f7e63969e27121b33 and is included since version 0.5.0. That's roughly 18 months back...
You may find documentation on this at https://github.com/promregator/promregator/blob/a0a6f2b01c02dbf386fe265bf32cfa633940a3fc/docs/documentation.md#label-rewriting.